element_calcium_imaging/imaging.py

```diff
@@ -1024,48 +1015,116 @@ class ActivityExtractionMethod(dj.Lookup):
+
+        if q_param:  # If the specified param-set already exists
+            pname = q_param.fetch1("activity_extraction_paramset_idx")
+            if (
+                pname == activity_extraction_paramset_idx
+            ):  # If the existed set has the same name: job done
+                return
+            else:  # If not same name: human error, trying to add the same paramset with different name
+                raise dj.DataJointError(
+                    "The specified param-set already exists - name: {}".format(pname)
+                )
+        else:
+            cls.insert1(param_dict)
 
 
 @schema
 class Activity(dj.Computed):
     definition = """  # inferred neural activity from fluorescence trace - e.g. dff, spikes
     -> Fluorescence
-    -> ActivityExtractionMethod
+    -> ActivityExtractionParamSet
     """
 
     class Trace(dj.Part):
-        definition = """  #
+        definition = """
         -> master
         -> Fluorescence.Trace
+        activity_type: varchar(16)  # e.g. dF/F, calcium-event
         ---
-        activity_trace: longblob  # 
+        activity_trace: longblob
         """
 
     @property
     def key_source(self):
+        ks = Fluorescence * ActivityExtractionParamSet
         suite2p_key_source = (
-            Fluorescence
-            * ActivityExtractionMethod
-            * ProcessingParamSet.proj("processing_method")
-            & 'processing_method = "suite2p"'
-            & 'extraction_method LIKE "suite2p%"'
+            ks
+            & (
+                Fluorescence
+                * ActivityExtractionMethod
+                * ActivityExtractionParamSet
+                * ProcessingParamSet.proj("processing_method")
+                & 'processing_method = "suite2p"'
+                & 'extraction_method = "suite2p"'
+            ).proj()
         )
         caiman_key_source = (
-            Fluorescence
-            * ActivityExtractionMethod
-            * ProcessingParamSet.proj("processing_method")
-            & 'processing_method = "caiman"'
-            & 'extraction_method LIKE "caiman%"'
+            ks
+            & (
+                Fluorescence
+                * ActivityExtractionMethod
+                * ActivityExtractionParamSet
+                * ProcessingParamSet.proj("processing_method")
+                & 'processing_method = "caiman"'
+                & 'extraction_method = "caiman"'
+            ).proj()
+        )
+        fissa_key_source = (
+            ks
+            & (
+                Fluorescence
+                * ActivityExtractionMethod
+                * ActivityExtractionParamSet
+                * ProcessingParamSet.proj("processing_method")
+                & 'processing_method in ("suite2p")'
+                & 'extraction_method = "FISSA"'
+            ).proj()
+        )
+        return (
+            suite2p_key_source.proj()
+            + caiman_key_source.proj()
+            + fissa_key_source.proj()
```

Comment:
If I am understanding this correctly, this key source would in effect return `Fluorescence * ActivityExtractionParamSet`.  Alternatively, should we add an `ActivityTask` table?

Reply:
Hi @ttngu207, what are your thoughts on this suggestion?

Reply:
union `+` is a very inefficient query and should not be used.